### PR TITLE
generic config spec

### PR DESCRIPTION
**Pull request by jacobweinstock**
update config spec to be able to handle more than one provider

### CODE REVIEW

**Comment by bgentry-ntap** [comment]
I think this is where we need the provider interface that we can call for specific provider operations. I'm thinking bridge pattern where each engine has a provider that it can call for various provider operations. 

Also, not a big fan of 'types' package, IMO types should be defined in the package they are used.

File: pkg/config/types/types.go [hunk 1]
```diff
@@ -1,153 +1,59 @@
...
-	Subject  string `yaml:"Subject,omitempty" json:"subject,omitempty"`
-	BasePath string `yaml:"BasePath,omitempty" json:"basepath,omitempty"`
-	User     string `yaml:"User,omitempty" json:"user,omitempty"`
-	Token    string `yaml:"Token,omitempty" json:"token,omitempty"`
+	Provider        string          `yaml:"Provider" json:"provider"`
+	ProviderVsphere ProviderVsphere `yaml:"ProviderVsphere,omitempty" json:"providervsphere,omitempty"`
+	SSH             SSH             `yaml:"SSH" json:"ssh"`
+	LogFile         string          `yaml:"LogFile,omitempty" json:"logfile,omitempty"`
+	Addons          Addons          `yaml:"Addons,omitempty" json:"addons,omitempty"`
+	cluster         `yaml:",inline" json:",inline" mapstructure:",squash"`
```
Comment (j-griffith) [inline pkg/config/types/types.go:10]: Would this need to be exported?

**Comment by j-griffith** [comment]
> I think this is where we need the provider interface that we can call for specific provider operations. I'm thinking bridge pattern where each engine has a provider that it can call for various provider operations.
> 
> Also, not a big fan of 'types' package, IMO types should be defined in the package they are used.

I agree on the types package comment.  A provider interface does seem like a good direction here rather than dealing with two discrete options, try to tie them together and get some consistency.

**Comment by jacobweinstock** [comment]
> > I think this is where we need the provider interface that we can call for specific provider operations. I'm thinking bridge pattern where each engine has a provider that it can call for various provider operations.
> > Also, not a big fan of 'types' package, IMO types should be defined in the package they are used.
> 
> I agree on the types package comment. A provider interface does seem like a good direction here rather than dealing with two discrete options, try to tie them together and get some consistency.

same here, I agree. I'll need to PR the new pkg structure and resubmit this PR then.

**Comment by jacobweinstock** [comment]
duplicate of https://github.com/NetApp/cake/pull/31

**Pull request closed without merging**